File: apps/admin/src/types/payroll.ts
export type PayrollRunListItem = {
  id: number;
  period_start: string;
  period_end: string;
  status: string;
  notes: string | null;
  require_verified_attendance: boolean;
  created_at: string;
};

export type PayrollRunCreatePayload = {
  period_start: string;
  period_end: string;
  notes?: string | null;
  require_verified_attendance?: boolean;
};

export type PayrollRunCreateResponse = {
  success: boolean;
  message: string;
  data: {
    payroll_run_id: number;
    status: string;
  };
};

export type PayrollRunStatusUpdatePayload = {
  status: string;
};

export type PayrollRunStatusUpdateResponse = {
  success: boolean;
  message: string;
  data: {
    payroll_run_id: number;
    status: string;
  };
};

export type PayrollDeductionCreatePayload = {
  payroll_item_id: number;
  deduction_type: string;
  amount: number;
  reason?: string | null;
};

export type PayrollWarning = {
  code: string;
  message: string;
};

export type PayrollDeductionCreateResponse = {
  success: boolean;
  message: string;
  data: {
    deduction_id: number;
    payroll_item_id: number;
    total_deduction_amount: number;
    net_amount: number;
    warning: PayrollWarning | null;
  };
};

export type PayrollDeductionItem = {
  id: number;
  deduction_type: string;
  amount: number;
  reason: string | null;
};

export type PayrollItem = {
  id: number;
  assignment_id: number;
  worker_profile_id: number;
  gross_amount: number;
  total_deduction_amount: number;
  net_amount: number;
  attendance_days: number;
  half_days: number;
  absent_days: number;
  payment_status: string;
  platform_margin: number | null;
  is_stale: boolean;
  deductions: PayrollDeductionItem[];
};

export type RecalculatePayrollItemResponse = {
  success: boolean;
  message: string;
  data: PayrollItem;
};

export type PayrollRunDetailResponse = {
  success: boolean;
  message: string;
  data: {
    payroll_run: {
      id: number;
      period_start: string;
      period_end: string;
      status: string;
      notes: string | null;
      require_verified_attendance: boolean;
    };
    items: PayrollItem[];
  };
};

File: apps/backend/app/schemas/payroll.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.core.payroll_constants import DeductionType, PayrollRunStatus
from app.utils.validators import strip_optional_text


class PayrollRunCreateSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    period_start: date
    period_end: date
    notes: str | None = Field(default=None, max_length=1000)

    @field_validator("notes")
    @classmethod
    def clean_notes(cls, value: str | None) -> str | None:
        return strip_optional_text(value)

    @model_validator(mode="after")
    def validate_period(self):
        if self.period_end < self.period_start:
            raise ValueError("period_end must be on or after period_start")
        return self


class PayrollDeductionAddSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    payroll_item_id: int
    deduction_type: str
    amount: int = Field(ge=0)
    reason: str | None = Field(default=None, max_length=1000)

    @field_validator("deduction_type")
    @classmethod
    def validate_deduction_type(cls, value: str) -> str:
        normalized = value.strip().lower()
        if normalized not in {item.value for item in DeductionType}:
            raise ValueError("Invalid deduction type")
        return normalized

    @field_validator("reason")
    @classmethod
    def clean_reason(cls, value: str | None) -> str | None:
        return strip_optional_text(value)


class PayrollRunStatusUpdateSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: str

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        normalized = value.strip().lower()
        if normalized not in {item.value for item in PayrollRunStatus}:
            raise ValueError("Invalid payroll status")
        return normalized


class PayrollItemResponseSchema(BaseModel):
    id: int
    payroll_run_id: int
    assignment_id: int
    worker_profile_id: int
    gross_amount: int
    total_deduction_amount: int
    net_amount: int
    attendance_days: int
    half_days: int
    absent_days: int
    payment_status: str
